### PR TITLE
Filter xss when saving the title

### PR DESCRIPTION
**Pull request by jonathanyeong**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related Tickets & Documents
https://github.com/forem/forem/issues/2204

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by jonathanyeong** [comment]
@nickytonline Looks like other editors are having this issue! https://usolved.net/blog/post/ckeditor-converts-html-entities-to-angled-brackets

**Comment by jonathanyeong** [comment]
Here's an update @nickytonline. I found the other half of the issue - `sanitize_and_decode` is where it auto closes the tag. Presumably, now that we're escaping the bracket entities we can remove this call to `sanitize_and_decode`. 

I don't have the full context though, would love others to weigh in here!

Here's the fix in action:
Fix issue with auto complete

**Comment by nickytonline** [comment]
Hey @jonathanyeong! This slipped through the cracks. Going to pull this locally to test it out.

**Comment by nickytonline** [comment]
@jonathanyeong, just FYI, that I've updated your branch with the latest from main, and Elastic search is no longer a dependency, so you can remove it locally.

**Comment by citizen428** [comment]
Hi @jonathanyeong, thanks for your contribution! Are you still planning on addressing Nick's feedback above?

**Comment by jonathanyeong** [comment]
Here's a video showing the latest changes.

https://user-images.githubusercontent.com/3861088/122710765-cb300480-d215-11eb-8585-35fa09820eb5.mp4

There's a flash of the original title before JS kicks in. I can make a change to [_v2_article_form](https://github.com/forem/forem/pull/12789/files#diff-6f417f63e88e9b7e4b28f32d77d1d6ed33fd8d0500c7494e29fdc31bfc7b482bR93). To run `sanitize_and_decode` as well. Thoughts?

**Comment by jonathanyeong** [comment]
> Hi @jonathanyeong, thanks for your contribution! Are you still planning on addressing Nick's feedback above?

Sorry @citizen428 and @nickytonline! I've been super absent. Happy to continue with this issue - made some changes addressing Nick's feedback.

File: app/views/articles/_v2_form.html.erb [hunk 1]
```diff
@@ -37,7 +37,7 @@
 </style>
 
 <main id="main-content"
-  data-article="<%= article.to_json(only: %i[id title cached_tag_list published body_markdown main_image organization_id canonical_url updated_at], methods: %i[series all_series]) %>"
+  data-article="<%= article.to_json(only: %i[id title cached_tag_list published body_markdown main_image organization_id canonical_url updated_at], methods: %i[series all_series decoded_title]) %>"
```
Comment (jonathanyeong) [inline app/views/articles/_v2_form.html.erb:40]: I introduced `decoded_title` which only decodes the HTML entities. Sanitization happens with FilterXSS, and also sanitize_and_decode when displaying the article.

**Comment by rhymes** [comment]
Hi @jonathanyeong, I merged `main` in to fix a conflict that arose when #13365 was merged in earlier, is there anything else to do @nickytonline in this PR?

File: app/views/articles/_v2_form.html.erb (hunk 1 above)
Comment (nickytonline) [reply app/views/articles/_v2_form.html.erb:40]: Refactor (blocking):

The issue is not completely fixed.

If a post is edited, the title is correct in edit/preview mode.

image

image

If the post is saved as a draft, the script tags are stripped due to XSS vulnerabilities.

image

Publishing a post is fine.

image

One other thing to note is the document.title is not decoded when published. In draft mode it's fine though.

image

Comment (nickytonline) [reply app/views/articles/_v2_form.html.erb:40]: @rhymes, are you OK with the `decoded_title` approach? I wonder if the title returned instead should always be decoded.

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: app/views/articles/_v2_form.html.erb (hunk 1 above)
Comment (rhymes) [reply app/views/articles/_v2_form.html.erb:40]: I'm okay either way! It seems like we use both in the component...

**Comment by jonathanyeong** [comment]
Woah that seems funky! Thanks for testing this out @nickytonline! I'll work on this as soon as I get a chance. Are we fine keeping the `decoded_title` approach? cc/ @rhymes

**Comment by citizen428** [comment]
Hi @jonathanyeong 👋 I just wanted to check in to see if you're still working on this PR?

**Comment by jonathanyeong** [comment]
Oh sorry! I've been a bit busy. I can keep working on this, but if this is critical then I'll happily let someone else take it over! 

Regarding to @nickytonline's comments, I tracked down where the preview title is being called. It seems to be rendered in the [articles/show.html.erb](https://github.com/forem/forem/blob/main/app/views/articles/show.html.erb#L118) template. Do you know when that is being used versus when the [Preview component](thttps://github.com/forem/forem/blob/main/app/views/articles/show.html.erb#L118) is being used?

**Comment by maestromac** [comment]
I went ahead and take over this with a different solution. Thank you very much for the hard work you put in this PR @jonathanyeong .

**Pull request closed without merging**